sklearn/utils/_threaded_libs.py

```diff
@@ -0,0 +1,433 @@
+    the supported C-libraries. This function works for libraries that are
+    already loaded in the interpreter and can be changed dynamically.
+
+    Parameters
+    ----------
+    limits : int or dict, (default=None)
+        Maximum number of thread that can be used in thread pools
+
+        If int, sets the maximum number of thread to `limits` for each C-lib
+        selected by `subset`.
+
+        If dict(supported_libraries: max_threads), sets a custom maximum number
+        of thread for each C-lib.
+
+        If None, does not do anything.
+
+    subset : string or None, optional (default="all")
+        Subset of C-libs to limit. Used only if `limits` is an int
+
+        "all" : limit all supported C-libs.
+
+        "blas" : limit only BLAS supported C-libs.
+
+        "openmp" : limit only OpenMP supported C-libs. It can affect the number
+                   of threads used by the BLAS C-libs if they rely on OpenMP.
+
+    reload_clib : bool, (default=False)
+        If `reload_clib` is `True`, first loop through the loaded libraries to
+        ensure that this function is called on all available libraries.
+
+    Returns
+    -------
+    dynamic_threadpool_size : dict
+        contains pairs `('clib': boolean)` which are True if `clib` have been
+        found and can be used to scale the maximal number of threads
+        dynamically.
+    """
+    wrapper = _get_wrapper(reload_clib)
+    return wrapper.set_thread_limits(limits, subset)
+
+
+def get_thread_limits(reload_clib=True):
+    """Return maximal thread number for threadpools in supported C-lib
+
+    Parameters
+    ----------
+    reload_clib : bool, (default=True)
+        If `reload_clib` is `True`, first loop through the loaded libraries to
+        ensure that this function is called on all available libraries.
+
+    Returns
+    -------
+    limits : dict
+        Contains the maximal number of threads that can be used in supported
+        libraries or None when the library is not available. The key of the
+        dictionary are "openmp_gnu", "openmp_intel", "openmp_win32",
+        "openmp_llvm", "openblas", "mkl" and "mkl_win32".
+    """
+    wrapper = _get_wrapper(reload_clib)
+    return wrapper.get_thread_limits()
+
+
+class thread_pool_limits:
+    """Change the default number of threads used in thread pools.
+
+    This class can be used either as a function (the construction of this
+    object limits the number of threads) or as a context manager, in a `with`
+    block.
+
+    Parameters
+    ----------
+    limits : int or dict, (default=None)
+        Maximum number of thread that can be used in thread pools
+
+        If int, sets the maximum number of thread to `limits` for each C-lib
+        selected by `subset`.
+
+        If dict(supported_libraries: max_threads), sets a custom maximum number
+        of thread for each C-lib.
+
+        If None, does not do anything.
+
+    subset : string or None, optional (default="all")
```

Comment:
Isn't this redundant with the `limits` parameter as a dict?
It's not clear to me why one would prefer one over the other.

Also the accepted keys of the dict for `limits` should be documented

Reply:
This is not redundant as the `dict` for limit permits to select specific implementation (for instance you can have two different openblas for `numpy` and `scipy`) while the `subset` (rename `user_api` in the loky version) permit to set globally, for all library implementing the BLAS or openmp protocol some limits.